Nederlandse spoorwegen

b10m opened the pull request:
## Description:

New sensor to pick up Dutch train timetable information.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3773 
## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
- platform: nederlandse_spoorwegen
    email: you@example.com
    password: !secret ns_password
    routes:
      - name: Rotterdam-Amsterdam
        from: Rtd
        to: Asd
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

b10m committed:
Add Nederlandse Spoorwegen sensor

b10m committed:
Remove unused 'delay' option.

b10m committed:
Merge branch 'dev' into nederlandse_spoorwegen

b10m committed:
Merge branch 'dev' into nederlandse_spoorwegen

fabaff commented on homeassistant/components/sensor/nederlandse_spoorwegen.py:
```diff
@@ -0,0 +1,136 @@
+CONF_TO = 'to'
+CONF_VIA = 'via'
+CONF_EMAIL = 'email'
+CONF_PASSWORD = 'password'
```
Please check the available constants in `const.py` and use them this.

b10m replied:
Right, easy. Will adjust.

fabaff commented on homeassistant/components/sensor/nederlandse_spoorwegen.py:
```diff
@@ -0,0 +1,136 @@
+    """Setup the departure sensor."""
+    import ns_api
+    nsapi = ns_api.NSAPI(
+        config.get(CONF_EMAIL), config.get(CONF_PASSWORD))
```
I assume that there will be an exception if the credentials are wrong. You need to catch that and make the platform setup fails.

b10m replied:
At this stage, it's merely storing the credentials. But I see your point that we should validate that before we add sensors (devices) that keep failing. I'm patching the code now to retrieve all valid station names at setup_platform level to see if the credentials are fine. Bonus points: this way I can also validate the station input by users.

fabaff replied:
Sounds like a smart way to do it.

fabaff commented on homeassistant/components/sensor/nederlandse_spoorwegen.py:
```diff
@@ -0,0 +1,136 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_EMAIL): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_DEPARTURES): [{
```
Please make sure that it's a [list](https://home-assistant.io/developers/development_validation/#lists).

fabaff commented on homeassistant/components/sensor/nederlandse_spoorwegen.py:
```diff
@@ -0,0 +1,136 @@
+            NSDepartureSensor(
+                nsapi, departure.get(CONF_NAME), departure.get(CONF_FROM),
+                departure.get(CONF_TO), departure.get(CONF_VIA)))
+    add_devices(sensors)
```
Add [`True`](https://home-assistant.io/developers/code_review_platform/#5-entity) as a second parameter.

fabaff commented on homeassistant/components/sensor/nederlandse_spoorwegen.py:
```diff
@@ -0,0 +1,136 @@
+        }
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
```
According to your configuration you allow multiple entries. I don't know how the `nsapi` module works but usually it's smarter to use a data object to avoid multiple requests.

b10m replied:
Not entirely sure how to handle this. The API queries a single route (from - (via?) - to) and returns data. You can't lookup multiple rounds in one call, so I guess I'll leave this as is for now.

b10m committed:
Apply requested code reviewed changes

- use constants from const.py
- ensure the configuration is a list
- verify credentials upon platform setup
- verify station input
- add True as second param to add_device
- only call add_device if at least 1 sensor was setup

b10m commented:
@fabaff Requested changes applied and committed.

fabaff reviewed (approved):
Thanks :bird:
